R/statistics.R

```diff
@@ -691,3 +691,247 @@ SecretarySampler <- R6::R6Class("SecretarySampler", public = list(
+ExploreFristNBandit <- R6::R6Class(
+    "ExploreFirstNBandit",
+    inherit = MultiArmedBandit,
+    public = list(
+        #' @description Creates a new \code{ExploreFirstNBandit} streamer
+        #' object.
+        #'
+        #' @param K the number of arms
+        #' @param T number of rounds / draws
+        #' @param N number of draws in the exploration phase, if not specified
+        #' the optimal value for N = (T/K) ^ (2/3) * log(T) ^ (1/3) is chosen.
+        #'
+        #' @examples
+        #' Bandit <- ExploreFristNBandit$new(3, 20, 2)
+        #'
+        #' @return The new \code{ExploreFristNBandit} (invisibly)
+        initialize = function(K, T, N = NULL) {
+            super$initialize(K, T)
+            private$state <- "exploration"
+            if (is.null(N)) {
+                N <- floor((T / K) ^ (2 / 3) * log(T) ^ (1 / 3))
+            }
+            if (T < N * K) {
+                stop(sprintf(
+                    "More draws are required for the values of
+                         N=%s and K=%s",
+                    N,
+                    K
+                ))
+            }
+            private$N <- N
+            private$which <- 1
+            invisible(self)
+        },
+        #' @description Update the \code{ExploreFristNBandit} streamer object.
+        #'
+        #' @param x a single draw from an arm
+        #' @param k index of the chosen arm
+        #'
+        #' @examples
+        #' Bandit$update(rnorm(1), 1)
+        #'
+        #' @return The updated code{ExploreFristNBandit} (invisibly)
+        update = function(x, k) {
+            super$update(x, k)
+            if (private$n_observed < K * N) {
+                private$which <- ceiling(private$n_observed / private$N)
+            } else if (private$n_observed == K * N) {
+                private$state <- "exploitation"
+                private$which <-
+                    which.max(private$get_state()$arm_means)
+            }
+        }
+    ),
+    private = list(N = NULL)
+)
+
+#' Create a streamer for the epsilon-greedy strategy in the multi-armed
+#' bandit problem.
+#'
+#' @description \code{EpsilonGreedyBandit}
+#'
+#' @docType class
+#'
+#' @examples
+#' means <- c(1, 2, 5)
+#' T <- 10
+#' K <- length(means)
+#' Bandit <- EpsilonGreedyBandit$new(K, T)
+#' for (t in seq(T)) {
+#'     # get which arm to choose
+#'     k <- Bandit$value$which
+#'     # update the streamer with a value drawn from the kth arm
+#'     x <- rnorm(1, mean = means[k])
+#'     Bandit$update(x, k)
+#' }
+#'
+#' @export
+#' @format An \code{\link{R6Class}} generator object
+EpsilonGreedyBandit <- R6::R6Class(
+    "EpsilonGreedyBandit",
+    inherit = MultiArmedBandit,
+    public = list(
+        #' @description Creates a new \code{EpsilonGreedyBandit} streamer
+        #' object.
+        #'
+        #' @param K the number of arms
+        #' @param T number of rounds / draws
+        #' @param epsilon vector of length T specifying the exploration
```

Comment:
Typically, epsilon greedy has a scalar epsilon held constant over the game (Sutton, R. S. & Barto, A. G. 1998 Reinforcement learning). I think something this general doesn't really need to exist. The closest thing is epsilon decreasing where epsilons form a geometric series.

Reply:
I relied on this: https://arxiv.org/pdf/1904.07272.pdf, see Algorithm 1.2,  epsilon can change in time. Perhaps we can leave it as it is and just add the possibility for epsilon to be a constant value.

Reply:
Fair enough about the source but for the purpose of online algorithms, specifying the whole epsilon doesn't make much sense. I think sticking with a fixed epsilon of a geometric series is more in-line with the packages vision.

Related to this, does it make sense that these games have finite time horizons? Shouldn't the point be that you can steam indefinitely?

Reply:
I agree + was also not convinced by the fine time T appearing in those algorithms. I will adjust them for infinite sampling.